### PR TITLE
Prompt to Download Missing Maps

### PR DESCRIPTION
**Pull request by DanVanAtta**
7 commits based on prior PRs

#### Prompt User to Download missing Maps
After joining a game where you do not have the map, and the game launches, today a stack trace is shown to the user with a message that the map could not be found. This stack trace kills the new process that would otherwise become the full blown game. I did not find a good way to replace that functionality, so we still have the stack trace, but we do at least have a dialog message that will help to download the map. So now instead of just showing the stack trace, we'll show a "yes/no" prompt asking the user if they would like to download the map they were missing. Screenshot below:

map_missing_diag

If they click yes, then the download dialog opens up and the missing map download begins immediately. Here is what that looks like after the download completes:

map_download_done

#### Move download Maps Button to GameSelectorPanel

Simply a move of the "Download Maps" button to below "load saved game" button.
- The Download maps button is now visible on game staging screens. I was looking at ways to have the button appear when we could detect that the player was missing the map (rather than waiting all the way to game launch). I did not find a clean way to do this, but found it's just useful to have the download maps button on the game staging screen. Players if they know they are missing the map, can download the map without leaving the staging screen. One can also check if they already have the map installed as well by opening the download map dialog and looking at the installed games tab.

Here is a screenshot showing the moved button:
main_menu_with_downloads_maps

[Review on Reviewable](https://reviewable.io/reviews/triplea-game/triplea/471)

### CODE REVIEW

**Comment by ron-murhammer** [comment]
@DanVanAtta - Looks like this needs rebased.

This might be a good time to revisit the location of the 'Game Options...' button as the screenshots here show how it was moved down during the previous PR.

**Comment by DanVanAtta** [comment]
@ron-murhammer , with respect to rebase, it's key we get some of the earlier PRs merged. The code updates that precede either are very core and a lot of things build on it (like Swing components), or alternatively I've already integrated the prior branches to avoid merge conflicts (IE: these changes conflict with those that are already in the queue, so it's based on those branches to integrate earlier)

So, if I were to resolve merge conflicts of this PR, I would have to redo the merge conflict resolutions of PRs#449, and #451. Having done so a number of times already.. it's really wasteful. If we merge the earlier PRs first, then I won't have to redo those merge conflicts anymore, and can just focus on the new changes presented in this PR.

**Comment by DanVanAtta** [comment]
Spacing fixed by putting the spacer element in grid row 10, rather than sharing grid row 9.

fixed_layout

**Comment by DanVanAtta** [comment]
Merge conflict resolved, but it was a really complicated one
This needs additional testing now.

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like the build is failing for this.

**Comment by DanVanAtta** [comment]
I'm quite sure it's still the case that you have to 'refresh games' or restart before new maps are available. Let's please revisit the messaging before we do the next release. I've cut this issue as  tracker/reminder: https://github.com/triplea-game/triplea/issues/527

File: src/games/strategy/engine/framework/map/download/InstallMapDialog.java [hunk 1]
```diff
@@ -53,37 +56,65 @@ public static Version getVersion(final File zipFile) {
...
+    showDownloadMapsWindow(parent, Optional.empty());
+  }
+
+  public static void showDownloadMapsWindow() {
+    showDownloadMapsWindow(null, Optional.empty());
+  }
+
 
-  public static void showDownloadMapsWindow(final Component parent, final List<DownloadFileDescription> games) {
+  private static void showDownloadMapsWindow(final Component parent, Optional<String> mapName) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/InstallMapDialog.java]: This line is perhaps the key to the change. When we create a 'download maps window' we give an optional map name which is a map that we wish to start downloading immediately.

**Comment by DanVanAtta** [comment]
ready for review

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like this needs rebased.

**Comment by DanVanAtta** [comment]
ready again.

**Comment by ron-murhammer** [comment]
@DanVanAtta - Functional feedback:

1. I get 2 popups when loading a single player game that I don't have the map for. Its pretty awkward. Probably should remove the old popup:
capture

2. After I download the map, it no longer is on the saved game and I have to choose 'Load Saved Game' again. Would be much better if it kept the save game active so I can just play then.

3. After a map downloads and you get the prompt, the actions are reversed as 'Yes' I want to download more maps closes the map downloader while 'No' doesn't.

4. I chose download map then selected a map and it downloaded but I got a stacktrace as well:
```
Exception: -1
java.util.ArrayList.elementData(ArrayList.java:418)
java.util.ArrayList.get(ArrayList.java:431)
games.strategy.engine.framework.map.download.InstallMapDialog.lambda$9(InstallMapDialog.java:237)
javax.swing.JList.fireSelectionValueChanged(JList.java:1796)
javax.swing.JList$ListSelectionHandler.valueChanged(JList.java:1810)
javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:184)
javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:164)
javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:211)
javax.swing.DefaultListSelectionModel.removeIndexInterval(DefaultListSelectionModel.java:677)
javax.swing.plaf.basic.BasicListUI$Handler.intervalRemoved(BasicListUI.java:2601)
javax.swing.AbstractListModel.fireIntervalRemoved(AbstractListModel.java:179)
javax.swing.DefaultListModel.removeElement(DefaultListModel.java:383)
games.strategy.engine.framework.map.download.InstallMapDialog.lambda$18(InstallMapDialog.java:383)
java.util.ArrayList.forEach(ArrayList.java:1249)
games.strategy.engine.framework.map.download.InstallMapDialog.lambda$15(InstallMapDialog.java:381)
javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
org.pushingpixels.substance.internal.utils.RolloverButtonListener.mouseReleased(RolloverButtonListener.java:124)
java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
java.awt.Component.processMouseEvent(Component.java:6535)
javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
java.awt.Component.processEvent(Component.java:6300)
java.awt.Container.processEvent(Container.java:2236)
java.awt.Component.dispatchEventImpl(Component.java:4891)
java.awt.Container.dispatchEventImpl(Container.java:2294)
java.awt.Component.dispatchEvent(Component.java:4713)
java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
java.awt.Container.dispatchEventImpl(Container.java:2280)
java.awt.Window.dispatchEventImpl(Window.java:2750)
java.awt.Component.dispatchEvent(Component.java:4713)
java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
java.awt.EventQueue.access$500(EventQueue.java:97)
java.awt.EventQueue$3.run(EventQueue.java:709)
java.awt.EventQueue$3.run(EventQueue.java:703)
java.security.AccessController.doPrivileged(Native Method)
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
java.awt.EventQueue$4.run(EventQueue.java:731)
java.awt.EventQueue$4.run(EventQueue.java:729)
java.security.AccessController.doPrivileged(Native Method)
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by DanVanAtta** [comment]
@ron-murhammer , responding to the 4 outstanding issues:
1. I'll remove the redundant prompt
2. I agree it would be better, but it's an existing problem, and I do not have capacity for the increased scope.
3. The swapped buttons is the politics popup 'fix'. The problem/fix do not belong in this PR
4. I'll take a look.. Either something new/missed, or bad merge resolution.

**Comment by DanVanAtta** [comment]
@ron-murhammer  pushed some commits to fix 1,3,4. 

For 2, thinking about it, there was a chance that the fix would have been easy. Turns out if we leave the labels, it could be misleading. Clicking play again will do nothing since there is no game data loaded.  Below is the code that nulls out the map/game labels:

`GameSelectorModel.java`:
```
    synchronized (this) {
      if (data == null) {
        m_gameName = m_gameRound = m_gameVersion = "-";
      } else {
```

**Comment by DanVanAtta** [comment]
It feels like a time warp with this branch, at least 3 things that had been fixed are broken again after incorrect auto-merge, or incorrect merge resolution.. (for the record, the merge conflicts in this one were *nasty*, so I'm not surprised there have been so many problems..)

1. download complete dialog should not show up after the download map window has been closed
broken again.

2. the install map size does not appear on initial download map window
broken again

3. the game options button spacing is broken again
fixed

**Comment by DanVanAtta** [comment]
Last patch fixes what is left of the outstanding issues:
- download complete prompt does not show up if you close the window, at least in the simple case. This PR does not really update that in one way or another, so it should be an existing bug in the system if possible. None too surprising since the map download progress bar was merged not that long ago.

**Comment by DanVanAtta** [comment]
ready again for review.

**Comment by ron-murhammer** [comment]
@DanVanAtta - This still doesn't seem to be working properly. I load a saved game that I don't have the map downloaded for. Then I choose 'start local game' then 'play'. I get the following error in addition to the new download map popup:

```
java.lang.IllegalStateException: Could not find file folder or zip for map: Greyhawk_Wars
Please DOWNLOAD THIS MAP if you do not have it.
If you are making a map or mod, make sure the mapName property within the xml game file exactly matches the map zip or folder name.


	at games.strategy.triplea.ResourceLoader.getMapResourceLoader(ResourceLoader.java:50)
	at games.strategy.triplea.ui.UIContext.internalSetMapDir(UIContext.java:67)
	at games.strategy.triplea.ui.AbstractUIContext.setDefaultMapDir(AbstractUIContext.java:123)
	at games.strategy.triplea.ui.TripleAFrame.<init>(TripleAFrame.java:225)
	at games.strategy.triplea.TripleA$1.run(TripleA.java:135)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by DanVanAtta** [comment]
The error message I don't think is really possible to get rid without very significant architectural changes. Here is the reasoning:
- We use an exception for control flow when a map is missing
- IN the single player case we catch the exception and the game is not loaded
- But, in the multiplayer case we launch a brand new game process. The exception is important there since it kills the new game and exits the user back to the lobby

So, we could swallow the exception, but if there is any other problem with the game launch, then we'll swallow the real problem. Fixing the error message in the background requires *much* more significant architectural changes. Adding in some hacks to fix it for this case is really not worth it, and we would likely wind up with inconsistent and/or buggy error handling.

**Comment by ron-murhammer** [comment]
@DanVanAtta I think we need to address this before the PR is merged as its a pretty serious issue. The functionality of loading a save game that you don't have the map for is better without this code than with it as at least you didn't get a stack trace. I'm fine pushing this off until after the next release.

**Comment by DanVanAtta** [comment]
I don't think I'm a fan of that plan:
- this is an important feature if we reduce the bundled map set. The primary use case of where you hit this is joining multiplayer games where you do not have the map. The missing save game is a much more remote case, which I would just as well leave unaddressed (in the one case you hit the issue by clicking around in the game, the second you have to actually do something to get and load the save, in which case it's much more reasonable for a person to hit an error when they are missing the map).
- the merge conflicts with this change were terrible. I'd rather integrate the code now unless there is a problem with it.
- the fixing of the stack trace is going beyond the initial scope of this feature. That would be fixing an additional problem that already existed with the game
- I do not want to sign up for fixing the stack trace, since that would likely involve re-writing how games are launched. I simply am not volunteering to do that.

**Comment by ron-murhammer** [comment]
@DanVanAtta - Responses:
1. It is probably needed if we are going to reduce bundled maps. Primary use case is 'multiplayer' as long as you are included PBF/PBEM as well. Missing map when loading a save game happens all the time in PBF/PBEM as well as lobby.
2. I believe our basic criteria is to make sure all PRs 'improve' the game and in this case I think the experience is actually worse and there is a problem with it. Essentially the happy path is broken here.
3. Disagree as we can't have this feature if we have stacktraces popping up in the basic happy path.
4. I haven't looked at what it'll take to resolve the stack trace but I definitely think it needs addressed in some manner.

**Comment by DanVanAtta** [comment]
When you download the game, and click the first button you see, it'll be join lobby. If you then join the first game, you'll be hit with the unable to find map. I recall when it was *just* a stack trace and didn't even tell you to download the map. Saving people to find and download the correct map is big.

**Comment by DanVanAtta** [comment]
So the comparison is a stack trace, vs stack trace + popup.
In the first case you have to read and know what the error message means. If you are not even sure how to download maps yet, it's not that fun. Then you have to find the right map to download, which is not trivial since they come bundled and so the name may not match. in the second you see a bad error message.

I am most concerned about further merge conflicts. This has already had 3 sets of merge conflicts, and the last set was *nasty* to resolve and involved re-writing large chunks of code.  I would like someone else's opinion that the overall UX is worse or not. @gaborbernat  or @veqryn , could you chime in on what you think? Whether it's better to just have a stack trace that says "you don't have this map, please DOWNLOAD it", or if it's better to have the same stack trace + a pop up that asks if you want to DL it?

**Comment by ron-murhammer** [comment]
@DanVanAtta - In the PBF/PBEM/singleplayer, the comparison is popup telling you to go download the map (current) vs stacktrace and popup giving you the option to download the map. The first is much more user friendly as you don't have stacktraces popping up and just click on download maps.

I think you are only comparing the 'lobby' experience. Please spend a few minutes testing a save game on 1.8.0.9 on local or PBF/PBEM game.

**Comment by DanVanAtta** [comment]
I can re-hack something to remove the stack trace in just the one case... stay tuned.

**Comment by DanVanAtta** [comment]
Patched, single player case no longer has the stack trace :cake:

**Comment by DanVanAtta** [comment]
@ron-murhammer  - ball is in your court. 

Net effect of this PR is the single player map-missing pop-up (LocalLauncher.java)  has been replaced, meanwhile we've added a pop-up to the multiplayer case that allows for a direct download. There is still the same error message there that happens. IMO the error message that says pls Download, and  a pop-up that asks me if I would like to download, more or less are okay together. I'd rather have the game pre-emptively tell me that I've loaded a map I cannot play, or that I've joined a game where I do not have the map.

**Comment by DanVanAtta** [comment]
@ron-murhammer , will merge this soon. All issues have been addressed in this PR. I would like to call out the cost of delaying integration. Also, PR reviews generally should be  'review' rather than 'gatekeeping'. If we continue down the gatekeeping route, reviews need to be more responsive. Said in another way, if one person takes a vacation and can no longer review, we need to be sure that all progress in the project does not stop, which is effectively where we are today (not as bad now that Gabor can merge, but we're not far from that situation).

**Comment by ron-murhammer** [comment]
@DanVanAtta Alright. I'll try to review this today. Haven't had much time this past week.

File: src/games/strategy/common/swing/SwingComponents.java [hunk 2]
```diff
@@ -30,12 +32,15 @@
...
 import javax.swing.border.EmptyBorder;
 
+import com.google.common.collect.Sets;
 import games.strategy.engine.framework.map.download.DownloadFileDescription;
 import games.strategy.net.DesktopUtilityBrowserLauncher;
 import games.strategy.triplea.UrlConstants;
 
 public class SwingComponents {
 
+  private static Set<String> visiblePrompts = Sets.newHashSet();
```
Comment (ron-murhammer) [inline src/games/strategy/common/swing/SwingComponents.java]: Better to use new HashSet<>();

File: src/games/strategy/common/swing/SwingComponents.java [hunk 3]
```diff
@@ -63,6 +68,10 @@ public static JPanel gridPanel(int rows, int cols) {
     return panel;
   }
 
+  public static JButton newJButton(String title, String toolTip, Runnable actionListener) {
+    return newJButton(title, toolTip, SwingAction.of(e->actionListener.run()));
```
Comment (ron-murhammer) [inline src/games/strategy/common/swing/SwingComponents.java]: Nitpick but adding spaces in the lambda expression would help.

File: src/games/strategy/common/ui/BasicGameMenuBar.java [hunk 4]
```diff
@@ -172,10 +171,8 @@ protected void addGameNotesMenu(final JMenu parentMenu) {
     if (notesProperty != null && notesProperty.trim().length() != 0) {
       final String notes = LocalizeHTML.localizeImgLinksInHTML(notesProperty.trim());
       m_gameNotesPane = new SoftJEditorPane(notes);
-      parentMenu.add(SwingAction.of("Game Notes...", e -> {
-        SwingUtilities.invokeLater(new Runnable() {
-          @Override
-          public void run() {
+      parentMenu.add(SwingAction.of("Game Notes...", e ->
+          SwingUtilities.invokeLater(() -> {
```
Comment (ron-murhammer) [inline src/games/strategy/common/ui/BasicGameMenuBar.java]: This is a little hard to follow with the lambda chaining. Might be better to split it up some.

File: src/games/strategy/engine/framework/map/download/DownloadCoordinator.java [hunk 5]
```diff
@@ -42,18 +42,18 @@ private final Thread createDownloadFileStarterThread() {
...
-            };
             SwingComponents.promptUser("Map Downloads Completed",
-                "Map downloads are complete, would you like to continue downloading more maps?", confirmAction);
+                "Map downloads are complete, would you like to continue downloading more maps?", () -> {}, // do nothing, keep the window open
+                () -> { // user clicked no, close the window
+                  if (downloadCompleteAction.isPresent()) {
+                    downloadCompleteAction.get().run();
+                  }
+                }
+            );
```
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/map/download/DownloadCoordinator.java]: This is kind of tough to follow. Might be better to define a variable for each action instead of defining everything in line.

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java [hunk 6]
```diff
@@ -49,20 +49,8 @@ protected void launchInNewThread(final Component parent) {
...
-        error = error.getCause();
-      }
-      final String message = error.getMessage();
-      m_gameLoadingWindow.doneWait();
-      SwingUtilities.invokeLater(new Runnable() {
-        @Override
-        public void run() {
-          JOptionPane.showMessageDialog(null, message, "Warning", JOptionPane.WARNING_MESSAGE);
-        }
-      });
```
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: What message was this showing? Does this potentially affect other areas outside of map download?

File: src/games/strategy/common/swing/SwingComponents.java [hunk 7]
```diff
@@ -91,17 +100,31 @@ public static void promptUser(final String title, final String message, final Ru
...
-          JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
-      boolean result = response == JOptionPane.YES_OPTION;
-
-      if (result) {
-        confirmedAction.run();
-      } else {
-        cancelAction.run();
+
+    boolean showMessage = false;
+    synchronized (visiblePrompts) {
```
Comment (ron-murhammer) [inline src/games/strategy/common/swing/SwingComponents.java]: Why is this synchronized needed? Is it to make sure we don't open the same dialog twice?

**Comment by ron-murhammer** [comment]
@DanVanAtta - Functionality seems good now. I glanced through the code and added a few comments.

File: src/games/strategy/common/swing/SwingComponents.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/common/swing/SwingComponents.java]: Why is that better?

File: src/games/strategy/common/swing/SwingComponents.java (hunk 7 above)
Comment (DanVanAtta) [inline src/games/strategy/common/swing/SwingComponents.java]: Swing actions are launched on independent threads, so to prevent the unlikely race condition.

File: src/games/strategy/common/ui/BasicGameMenuBar.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/common/ui/BasicGameMenuBar.java]: We're just replacing the Runnable anonymous inner class.

File: src/games/strategy/engine/framework/map/download/DownloadCoordinator.java (hunk 5 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/DownloadCoordinator.java]: The lib function is meant to be easy enough to follow. I'd rather fix the API, or let this one be a 'know your libs'. I do not want to have to label the params every time we use this function.

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: This was the 'duplicate' single player "you are missing the map" message.

File: src/games/strategy/common/swing/SwingComponents.java (hunk 2 above)
Comment (ron-murhammer) [inline src/games/strategy/common/swing/SwingComponents.java]: Because its native Java and just as straightforward/concise. No reason to use a library for that.

From the Guava docs, "Note for Java 7 and later: this method is now unnecessary and should be treated as deprecated. Instead, use the HashSet constructor directly, taking advantage of the new "diamond" syntax." https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Sets.html#newHashSet()

File: src/games/strategy/common/ui/BasicGameMenuBar.java (hunk 4 above)
Comment (ron-murhammer) [inline src/games/strategy/common/ui/BasicGameMenuBar.java]: Right. It was already difficult to follow before any of the lambda refactor and now is probably even harder to follow. If you take a step back and look at addGameNotesMenu method its pretty brutal. In particular these couple of lines chained together are very difficult to determine what's going on:
```
      parentMenu.add(SwingAction.of("Game Notes...", e ->
          SwingUtilities.invokeLater(() -> {
            final JEditorPane pane = m_gameNotesPane.getComponent();
```

File: src/games/strategy/engine/framework/map/download/DownloadCoordinator.java (hunk 5 above)
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/map/download/DownloadCoordinator.java]: I think inline is probably ok for single line Runnables but don't you think this block of code is pretty rough?

```
if (shouldShowDownloadsFinishedPrompt()) {
            // TODO: we ignore the threading problem here, we may enter this code block multiple times.
            // It's a difficult situation to create for the user, and the worst case is multiple map download
            // complete dialogs appearing - which is not a big deal.
            downloadPromptAlreadyShown = true;
            SwingComponents.promptUser("Map Downloads Completed",
                "Map downloads are complete, would you like to continue downloading more maps?", () -> {}, // do nothing, keep the window open
                () -> { // user clicked no, close the window
                  if (downloadCompleteAction.isPresent()) {
                    downloadCompleteAction.get().run();
                  }
                }
            );
          }
```

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java (hunk 6 above)
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: And do we know if this potentially was used in other situations? Just worried we could have gotten rid of a bunch of other useful dialogs as this is a pretty high level class.

File: src/games/strategy/common/swing/SwingComponents.java (hunk 7 above)
Comment (ron-murhammer) [inline src/games/strategy/common/swing/SwingComponents.java]: Are you sure this would prevent that? Can you explain the scenario?

Reading through it, I'm not sure that it handles all cases and might to need expand the synchronized scope. Couldn't you have one thread run through the sync block then another flip the showMessage flag to false which would stomp on showing the first prompt?

Potentially even worse is that message wouldn't be removed in the second block and then its already in the list so couldn't be re-added in the sync block.

Comment (DanVanAtta) [inline src/games/strategy/common/swing/SwingComponents.java]: If this action were launched from a button, double clicking it could have the threads race each other. Worst case a confirmation dialog pops up twice, which is not the end of the world either.


> Couldn't you have one thread run through the sync block then another flip the showMessage flag to false which would stomp on showing the first prompt?

No, the showMessage flag is method local, each thread has its own copy. If two threads were to pile up on the synchronized block, they would both have `false` for the value of `showMessage` Then when the next thread is allowed in the synchronized block, they will see that `visiblePromps` contains its message and will then fall through.

File: src/games/strategy/common/ui/BasicGameMenuBar.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/common/ui/BasicGameMenuBar.java]: The anonymous inner class version is more difficult IMO. The really brutal stuff is why a SwingAction is being adding to a parentMenu, doesn't make sense. Swing names and conventions could have been better, pretty amazingly old stuff.

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: Changed the IllegalState to a MapNotFoundException, now we can be sure the exact case where we throw MapNotfoundException will be caught here.

**Comment by DanVanAtta** [comment]
Future PRs have merge conflicts with this, and we are getting issues/queries about problems which this PR fixes. Integrating code is useful in its own right. I'm happy to continue working on the outstanding items, but really think it is time these changes were integrated. This branch has lived for far too long.

**Comment by ron-murhammer** [comment]
@DanVanAtta Agree, I'll try to look it over this evening if not then definitely this weekend. To be fair, it has been on your plate for almost 2 weeks but yeah let's work to get it merged.

**Comment by DanVanAtta** [comment]
Unless you think there is anything wrong, IMO let's merge now and then continue to work on the nit picky type of stuff.

It would be interesting to understand better why this PR took more than 2 months. I was not happy resolving merge conflicts so many times, plus the last set were extremely time consuming to resolve.

File: src/games/strategy/common/ui/BasicGameMenuBar.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/common/ui/BasicGameMenuBar.java]: To be clear, I do not intend to change this, the other two comments have been addressed. So everything outstanding should now be addressed.

**Comment by ron-murhammer** [comment]
@DanVanAtta I'd like to at least look over your latest commits and do a quick functionality check. Don't have the time at this moment to do so.

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java (hunk 6 above)
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: So wouldn't it be best to still leave the old exception handling here as well in case it is hit in other situations. So first catch MapNotFoundException then catch the IllegalStateException?

**Comment by ron-murhammer** [comment]
@DanVanAtta Just one outstanding question on the exception handling change.

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: Then we would have a pop-up message with the illegal state message and no stack trace. Given the unknown circumstances under which that would occur, the error message alone would be a bit lacking.

Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: Also notice that 'Exception' and 'IllegalState' are handled the same way now. It seems likely the same really random error would hit either the IllegalState block, or Exception, who knows what the code would do, and so it doesn't seem to reason why they should be handled differently.

In the previous case it was obvious the IllegalStateException was hijacked for the map not found purpose, for example note the extremely generous detail on the error message. We do not have examples where an IllegalState is so well documented with the messaging alone.
